ssm_jax/lgssm/core.py

```diff
@@ -0,0 +1,274 @@
+import jax.numpy as jnp
+import jax.random as jr
+from jax import lax
+
+import chex
+
+from distrax import MultivariateNormalFullCovariance as MVN
+
+
+@chex.dataclass
+class LGSSMParams:
+    """Lightweight container for LGSSM parameters."""
+    initial_mean: chex.Array
+    initial_covariance: chex.Array
+    dynamics_matrix: chex.Array
+    dynamics_input_weights: chex.Array
+    dynamics_covariance: chex.Array
+    emission_matrix: chex.Array
+    emission_input_weights: chex.Array
+    emission_covariance: chex.Array
+    is_stationary: chex.Array = True
+
+    @property
+    def state_dim(self):
+        return self.emission_matrix.shape[-1]
+
+    @property
+    def emission_dim(self):
+        return self.emission_matrix.shape[-2]
+
+
+def sample(self, rng, params, num_timesteps, inputs):
+    """_summary_
+
+    Args:
+        rng (_type_): _description_
+        params (_type_): _description_
+        num_timesteps (_type_): _description_
+        inputs (_type_): _description_
+    """
+    def _step(carry, rng_and_t):
+        state = carry
+        rng, t = rng_and_t
+
+        # Shorthand: get parameters and inputs for time index t
+        get = lambda x: x[t] if x.ndim == 3 else x
```

Comment:
You could do this (so it also initializes to 0s if argument is missing):
```
def get(matrix_or_fn, t, nrows=None, ncols=None):
  if matrix_or_fn is None:
     return jnp.zeros((nrows, ncols))
  elif iscallable(matrixor_fn):
     return matrix_or_fn(t)
  else
    return matrix_or_fn
    ```

and then use
```
 A = get(params.dynamics_matrix, t, nstates, nstates)
 B = get(params.dynamics_input_weights, t, nstates, ninputs)
...
```

Reply:
I was thinking about something like this. One concern is that JAX won't let you jit a function that takes a function as input ("not a valid JAX type"), so I thought for these low-level interfaces I'd assume the inputs were always arrays. This entails a bit of memory inefficiency for nonstationary cases where you have to construct all the transition matrices in memory, but it's a simpler interface.